activation/nipost.go

```diff
@@ -247,46 +239,109 @@ func (nb *NIPostBuilder) submitPoetChallenges(ctx context.Context, challenge []b
-		if err != nil {
-			nb.log.With().Panic("failed to fetch membership for poet proof", log.Binary("challenge", challenge[:]))
+	}
+	return false
+}
+
+func (nb *NIPostBuilder) getProofWithRetry(ctx context.Context, client PoetProvingServiceClient, roundID string, retryInterval time.Duration) (*types.PoetProofMessage, error) {
+	for {
+		proof, err := client.GetProof(ctx, roundID)
+		switch {
+		case err == nil:
+			return proof, nil
+		case errors.Is(err, ErrUnavailable) || errors.Is(err, ErrNotFound):
+			nb.log.With().Info("Proof not found, retrying", log.Duration("interval", retryInterval))
+			select {
+			case <-ctx.Done():
+				nb.log.Info("Retrying is canceled")
+				return nil, ctx.Err()
+			case <-time.After(retryInterval):
+			}
+		default:
+			return nil, err
 		}
-		if !membership[*challenge] {
-			nb.log.With().Debug("poet proof membership doesn't contain the challenge", log.Binary("challenge", challenge[:]))
+	}
+}
+
+func (nb *NIPostBuilder) getBestProof(ctx context.Context, challenge *types.Hash32) types.PoetProofRef {
+	proofs := make(chan *types.PoetProofMessage, len(nb.state.PoetRequests))
+
+	var eg errgroup.Group
+	for _, r := range nb.state.PoetRequests {
+		logger := nb.log.WithFields(log.String("poet-id", hex.EncodeToString(r.PoetServiceID)), log.String("round", r.PoetRound.ID))
+		client := nb.getPoetClient(ctx, r.PoetServiceID)
+		if client == nil {
+			logger.Warning("Poet client not found")
 			continue
 		}
-		proof, err := nb.poetDB.GetProof(ref)
-		if err != nil {
-			nb.log.Panic("Inconsistent state of poetDB. Received poetProofRef which doesn't exist in poetDB.")
-		}
-		nb.log.With().Info("Got a new PoET proof", log.Uint64("leafCount", proof.LeafCount), log.Binary("ref", ref))
+		round := r.PoetRound.ID
+		// Time to wait before quering for the proof
+		// The additional second is an optimization to be nicer to poet
+		// and don't accidentially ask it to soon and have to retry.
+		waitTime := time.Until(r.PoetRound.End.IntoTime()) + time.Second
+		eg.Go(func() error {
+			logger.With().Info("Waiting till poet round end", log.Duration("wait time", waitTime))
+			select {
+			case <-ctx.Done():
+				logger.With().Info("Waiting interrupted", log.Err(ctx.Err()))
+				return ctx.Err()
+			case <-time.After(waitTime):
+			}
+			proof, err := nb.getProofWithRetry(ctx, client, round, time.Second)
+			if err != nil {
+				logger.With().Warning("Failed to get proof from Poet", log.Err(err))
+				return nil
+			}
+
+			if err := nb.poetDB.ValidateAndStore(ctx, proof); err != nil && !errors.Is(err, ErrObjectExists) {
+				logger.With().Warning("Failed to validate and store proof", log.Err(err), log.Object("proof", proof))
+				return nil
+			}
 
-		if bestProof == nil || bestProof.leafCount < proof.LeafCount {
-			bestProof = &poetProof{
-				ref:       ref,
-				leafCount: proof.LeafCount,
+			// We are interested only in proofs that we are members of
+			if !membersContain(proof.Members, challenge) {
```

Comment:
does this warrant a warning log? 
the challenge is expected to be in the members if we have successfully submit it to the service id. right?

Reply:
Yes. Hitting this condition would mean that something is wrong with the poet. Changed to a warning log. Thanks!